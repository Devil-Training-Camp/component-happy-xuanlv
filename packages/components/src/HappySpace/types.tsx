import type React from 'react';
import type { SizeType } from '../types';

export interface HappySpaceProps extends React.HTMLAttributes<HTMLElement> {
  /**
   * 间距大小
   */
  size?: SpaceSizeType | SpaceSizeType[];
  /**
   * 间距方向
   */
  direction?: 'vertical' | 'horizontal';
  /**
   * 对齐方式
   */
  align?: 'start' | 'end' | 'center' | 'baseline';
}

export type SpaceSizeType = SizeType | number;
