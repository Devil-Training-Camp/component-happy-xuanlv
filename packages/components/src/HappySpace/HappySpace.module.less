@import '../style/variable-default';

.happySpace {
  display: flex;
  margin-bottom: calc(var(--ux-space-vertical-size) * -1);

  > * {
    flex: none;
    margin-bottom: var(--ux-space-vertical-size);
  }
}

.horizontal {
  > :not(:last-child) {
    margin-right: var(--ux-space-horizontal-size);
  }
}

.vertical {
  flex-direction: column;
}
