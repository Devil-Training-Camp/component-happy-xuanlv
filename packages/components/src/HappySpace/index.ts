export { default as HappySpace } from './HappySpace';
