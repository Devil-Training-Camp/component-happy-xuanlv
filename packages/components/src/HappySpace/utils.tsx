import type { SpaceSizeType } from './types';

const spaceSize = {
  small: 8,
  middle: 16,
  large: 24,
};

export function getNumberSize(size: SpaceSizeType) {
  return typeof size === 'string' ? spaceSize[size] : size || 0;
}
