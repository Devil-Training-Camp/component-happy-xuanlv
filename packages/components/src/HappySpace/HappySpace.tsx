import React, { useMemo } from 'react';
import clsx from 'clsx';
import type { HappySpaceProps } from './types';
import { getNumberSize } from './utils';
import styles from './HappySpace.module.less';

function HappySpace(
  {
    size = 'small',
    direction = 'horizontal',
    align = direction === 'horizontal' ? 'center' : undefined,
    className,
    style,
    children,
    ...restProps
  }: HappySpaceProps,
  ref: React.ForwardedRef<HTMLDivElement>,
) {
  const [horizontalSize, verticalSize] = useMemo(
    () => (Array.isArray(size) ? size : [size, size]).map(getNumberSize),
    [size],
  );

  return (
    <div
      className={clsx(styles.happySpace, className, {
        [styles.horizontal]: direction === 'horizontal',
        [styles.vertical]: direction === 'vertical',
      })}
      style={
        {
          '--ux-space-horizontal-size': `${horizontalSize}px`,
          '--ux-space-vertical-size': `${verticalSize}px`,
          alignItems: align,
          ...style,
        } as React.CSSProperties
      }
      ref={ref}
      {...restProps}
    >
      {children}
    </div>
  );
}

export default React.forwardRef(HappySpace);
