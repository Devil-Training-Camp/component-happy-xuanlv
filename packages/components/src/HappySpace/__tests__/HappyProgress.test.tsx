import React from 'react';
import { describe, it } from 'vitest';
import HappySpace from '../HappySpace';
import mountTest from '../../../tests/shared/mountTest';
import snapshotTest from '../../../tests/shared/snapshotTest';

describe('HappySpace', () => {
  mountTest(<HappySpace />);

  it('正常渲染', async () => {
    snapshotTest(<HappySpace size="small" />);
    snapshotTest(<HappySpace size="middle" />);
    snapshotTest(<HappySpace size="large" />);
    snapshotTest(<HappySpace size={8} />);
    snapshotTest(<HappySpace size={[8, 12]} />);
    snapshotTest(<HappySpace direction="vertical" />);
  });
});
