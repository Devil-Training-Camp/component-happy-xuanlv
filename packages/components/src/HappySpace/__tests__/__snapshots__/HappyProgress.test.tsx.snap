// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HappySpace > 正常渲染 1`] = `
<div
  class="_happySpace_056a30 _horizontal_056a30"
  style="--ux-space-horizontal-size: 8px; --ux-space-vertical-size: 8px; align-items: center;"
/>
`;

exports[`HappySpace > 正常渲染 2`] = `
<div
  class="_happySpace_056a30 _horizontal_056a30"
  style="--ux-space-horizontal-size: 16px; --ux-space-vertical-size: 16px; align-items: center;"
/>
`;

exports[`HappySpace > 正常渲染 3`] = `
<div
  class="_happySpace_056a30 _horizontal_056a30"
  style="--ux-space-horizontal-size: 24px; --ux-space-vertical-size: 24px; align-items: center;"
/>
`;

exports[`HappySpace > 正常渲染 4`] = `
<div
  class="_happySpace_056a30 _horizontal_056a30"
  style="--ux-space-horizontal-size: 8px; --ux-space-vertical-size: 8px; align-items: center;"
/>
`;

exports[`HappySpace > 正常渲染 5`] = `
<div
  class="_happySpace_056a30 _horizontal_056a30"
  style="--ux-space-horizontal-size: 8px; --ux-space-vertical-size: 12px; align-items: center;"
/>
`;

exports[`HappySpace > 正常渲染 6`] = `
<div
  class="_happySpace_056a30 _vertical_056a30"
  style="--ux-space-horizontal-size: 8px; --ux-space-vertical-size: 8px;"
/>
`;
